service/switchboardapp/src/test/java/com/switchboard/app/DeviceControllerTest.java

```diff
@@ -0,0 +1,105 @@
+package com.switchboard.app;
+
+import com.switchboard.app.controller.DeviceController;
+import com.switchboard.app.dao.DeviceDaoImpl;
+import com.switchboard.app.domain.DeviceEntity;
+import com.switchboard.app.exceptions.DeviceAlreadyExistsException;
+import com.switchboard.app.exceptions.DeviceNotFoundException;
+import com.switchboard.app.fixture.DeviceFixture;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+import org.springframework.http.ResponseEntity;
+import org.springframework.mock.web.MockHttpServletRequest;
+import org.springframework.web.context.request.RequestContextHolder;
+import org.springframework.web.context.request.ServletRequestAttributes;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.*;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.mockito.Mockito.when;
+
+class DeviceControllerTest {
+    @InjectMocks
+    private DeviceController deviceController;
+
+    @Mock
+    private DeviceDaoImpl deviceService;
+
+    //stubbed DeviceEntity object
+    static private DeviceEntity device1, device2;
+    static private List<DeviceEntity> listOfDevices;
+
+    @BeforeAll
+    static void deviceFixture(){
+        device1 = DeviceFixture.getDevice1();
+        device2 = DeviceFixture.getDevice2();
+        listOfDevices = DeviceFixture.getListOfDevices();
+    }
+
+    @BeforeEach
+    void setup(){
+        MockitoAnnotations.initMocks(this);
+    }
+
+    @Test
+    final void testRetrieveAllDevices(){
+        when(deviceService.getDevices()).thenReturn(listOfDevices);
+
+        List allDevices = deviceController.retrieveAllDevices();
+
+        assertFalse(allDevices.isEmpty(),"allDevices list is empty."); //check if an empty list was returned
+        assertIterableEquals(listOfDevices, allDevices,"listOfDevices and allDevices lists are not equal."); //check both lists contents
+    }
+
+    //When a device is available in the DB
+    @Test
+    final void testRetrieveDevice(){
+        when(deviceService.findDevice("1")).thenReturn(java.util.Optional.of(device1));
+
+        DeviceEntity actualDevice = deviceController.retrieveDevice("1").getContent();
+
+        assertNotNull(actualDevice, "actualDevice object is null.");
+        assertEquals(device1, actualDevice, "expectedDevice and actualDevice objects are not equal.");
+    }
+
+    //When a device is unavailable in the DB
+    @Test
+    final void testRetrieveDeviceEmpty(){
+        assertThrows(DeviceNotFoundException.class, () -> {
```

Comment:
`DeviceNotFoundException.class` needs to change to `ExceptionType.DeviceNotFoundException.class`

If not changed, we will get a compile error on merge.